File: IntelliJ_Projects/src/StronglyConnectedComponents/StronglyConnectedComponents.java
package StronglyConnectedComponents;
import java.util.*;

public class StronglyConnectedComponents {
    public static List<List<Integer>> scc(int students, List<List<Integer>> knows) {
        List<Integer>[] graph = form_graph(knows, students);

        boolean[] used = new boolean[students];
        List<Integer> order = new ArrayList<>();
        for (int i = 0; i < students; i++) {
            if (!used[i])
                dfs(graph, used, order, i);
        }

        List<Integer>[] reverseGraph = new List[students];
        for (int i = 0; i < students; i++)
            reverseGraph[i] = new ArrayList<>();
        for (int i = 0; i < students; i++)
            for (int j : graph[i])
                reverseGraph[j].add(i);
        List<List<Integer>> components = new ArrayList<>();
        Arrays.fill(used, false);
        Collections.reverse(order);

        for (int u : order)
            if (!used[u]) {
                List<Integer> component = new ArrayList<>();
                dfs(reverseGraph, used, component, u);
                components.add(component);
            }

        return components;
    }

    static void dfs(List<Integer>[] graph, boolean[] used, List<Integer> res, int u) {
        used[u] = true;
        for (int v : graph[u])
            if (!used[v])
                dfs(graph, used, res, v);
        res.add(u);
    }

    static List<Integer>[] form_graph(List<List<Integer>> knows, int students) {
        List<Integer>[] g = new List[students];
        for (int i = 0; i < students; i++)
            g[i] = new ArrayList<>();
        for (List<Integer> u : knows) {
            g[u.get(0)].add(u.get(1));
        }

        return g;
    }
}

File: Python_code/StronglyConnectedComponents/StronglyConnectedComponents.py
from collections import defaultdict

class StronglyConnectedComponents:
    graph = defaultdict(list)

    def scc(self, students: int, knows: [[int]]) -> [[int]]:
        used = [False] * students
        order = []
        self.form_graph(knows)
        for i in range(students):
            if not used[i]:
                self.dfs(self.graph, used, order, i)
        
        reverse_Graph = []
        for i in range(students):
            temp = []
            for j in self.graph[i]:
                temp.append(j)
            reverse_Graph.append(temp)
        
        components = []
        used = [False] * students

        for v in order:
            if not used[v]:
                component = []
                self.dfs(reverse_Graph, used, component, v)
                components.append(component)
        return components

    def form_graph(self, knows):
        for u, v in knows:
            self.graph[u].append(v)

    def dfs(self, graph, used, res, v):
        used[v] = True
        for k in graph[v]:
            if not used[k]:
                self.dfs(graph, used, res, k)
        res.append(v)